MGMT-11752: enhancement doc: ai on zero worker hypershift (#4325)
openshift/assisted-service: docs/enhancements/ai-on-zero-worker-hypershift.md

---
title: ai-on-zero-worker-hypershift
authors:
  - "@slaviered"
creation-date: 2022-08-28
last-updated: yyyy-mm-dd
---
#  Installing the Infrastructure Operator on zero-node HyperShift (DRAFT)

# Summary
To support large numbers of HyperShift hosted control planes on a single management cluster, an architecture of nested control planes has been proposed.  The management cluster, which we refer to as L0, runs hypershift hosted control planes (L1). Spoke clusters are added to L1 control planes. The challenge is that we need to run the assisted-installer in L1 for adding workers to spoke clusters, but L1 doesn't have any nodes and therefore cannot run the installer directly.

The architecture for multiple hubs is described in the following diagram
<img src="https://github.com/xiangjingli/hyper-acm-deploy/blob/main/docs/images/acm-hosted-cluster-diagram.svg" width="30%" height="30%"/>

> The hosted control plane (L1) is responsible for managing the CR and CRDs, where as the actual pods are located in dedicated namespaces and managed by the hosting management cluster (L0).

Reviewer: > actual pods

Clarify actual pods of what

> The hosted control plane (L1) is responsible for managing the CR and CRDs, where as the actual pods are located in dedicated namespaces and managed by the hosting management cluster (L0).

Reviewer: >managing the CR and CRDs

Clarify which CRs and which CRDs? Assisted?

The deployment and cluster provisioning on such a setup is described in this [POC](https://github.com/javipolo/openshift-assisted-installer-tests/tree/main/ai-in-zero-worker-hypershift). For convinience, the architecure of the deployment is also pasted ![here:](../ai-on-hypershift.png).

# Motivation
## Goals

> - Automate the deployment of assisted service on zero-nodeworker hypershift clusters using the assisted infrastructure operator

Reviewer 2: ```suggestion
- Automate the deployment of assisted service on zero-node hypershift clusters using the assisted infrastructure operator
```

> - Support Infrastructure Operatorai installation on multiple namespaces.

Reviewer 2: ```suggestion
- Support Infrastructure Operator installation in multiple namespaces.
```

Reviewer 2: Also the goal isn't to install the infrastructure operator into multiple namespaces, right? It's to allow it to install multiple instances of the service. I think that might be what you're trying to say here, but I would make it more clear.

Author: Both fixed.

- Isolation: Each instance of assisted service controls the spoke cluster creation on its corresponding hypershift-based hub cluster
- Backward Compatiability: Support automatic update of the infrastructure operator and preserve the current behavior on non-hypershift installations.

# Proposal
## User Story

> 1. The user installs the assisted insfrastructure operator

Reviewer 2: These are not really user stories, right? This is more of an example flow.

For a user story, I'd expect something more like:

`As an administrator, I can install assisted service into several namespaces in order for each to watch resources in a hosted control plane`

Then maybe something about how a user should expect assisted service to behave when installing a cluster.

Reviewer 2: To be clear though, I think this list of events is useful but should be put somewhere else. Maybe in the implementation details sections.

Author: I changed the title to reflect that we are listing the installation steps from the user's perspective. It is a bit out of place in the Implementation Details section so let's keep it in a separate section.

2. The user creates a namespace on the management cluster

> 3. The user creates a CR ion that namespace and provides the target Hhypersshift kubeconfig

Reviewer 2: ```suggestion
3. The user creates a CR in that namespace and provides the target HyperShift kubeconfig
```

Also what CR?

4. The system deploys the sub-components required for provisioning clusters
5. The user creates the spoke cluster's deployment-related CRs (i.e. infraenv, aci) in the hyperhisft cluster
6. The spoke cluster is generated

## Implementation Details
## Deployment of AI components with hostedCluster (L1) and managementCluster (L0)
Productization of this POC requires some modification to the assisted-service server and the infrastructure operator that deploys and configures it. The main challenges in the transition from a single hub installed on a physical cluster to multiple hubs on a hypershift platform are:

1) Deployment of assisted-service and its subsystems is now done per namespace

> 2) Assisted-service controllers (i.e. infra-env controller etc.) should now work with L1's api-server

Reviewer 2: ```suggestion
2) Assisted-service controllers (i.e. infra-env controller etc.) should now watch resources using L1's api-server
```

3) Infrastructure operator should deploy and govern artifacts on both L1 and L0

> 4) Webhook deployment is spanned over 2 api-servers in the hypershift mode and is now much more elaborated to deploy

Reviewer 2: ```suggestion
4) Webhook deployment is spanned over 2 api-servers in the hypershift mode and is now much more complicated to deploy
```

> ### Running on multiple namespaces

Reviewer 2: ```suggestion
### Running in multiple namespaces
```

Currently, the infrastructure operator is defined as a cluster scoped resource, and the namespace in which it is deployed (usually assisted-installer) is hard wired in the operator's code. Several approaches were considered for the solution:

<b>AgentServiceConfig CRD shall be converted to <i>Namespaced</i> scope (Rejected)</b>:
Reusing AgentServiceConfig while converting its scope was rejected because the operator can not update existing CRs (scope is an immutable field)<br>
<b>Deploy multiple AgentServiceConfig CRs on a cluster scope.</b>:
The target namespace will be part of the Spec. When the Spec does not contain the target namespace and other hypershift-related fields, then the deployment is assumed to behave as usual.

> This option was rejected because it requires too much behavioral branching in the current implementation and requires taking care of edge cases such as detecting when the user deployed multiple CRDs in a non-hypershift environment and rejecting the such option. Also, having the namespace field in the Spec is a hack.

Reviewer 2: Should these maybe be moved to the "Alternatives" section?

<b>Selected Solution: New CRD HypershiftServiceConfig on a namespace scope</b>
- Separation of intent
- Natural namespacing

### Assisted-service controllers working with hostedCluster (L1) control plane
From the user point of view, they define the cluster installation CRs on one of the namespaces in the hypershift cluster, and the assisted service picks them up and spins the cluster. To achieve that, we need to:
1. Have all cluster deployment CRDs defined on L1 
2. Assisted service's controllers should work with L1 api-server (the hosted cluster) exclusively, even though the pods on which the controllers are running are hosted on L0 (the management cluster).

#### CRD deployment
This task, which is currently done by the OLM on the management cluster, should now carried by the operator.

> There are also 2 options for finding which CRDs we need to deploy:

Reviewer 2: IMO the ideal option would be for whatever is creating the L1 cluster to ensure the required CRDs are synced.
This would move the responsibility of creating CRD resources back to the operator install layer (where it currently is in a single-control-plane use case).

This still needs us to sync with the team putting this all together though.

1. Build mechanism takes the CRDs from the deploy yaml and stores them on the filesystem of the operator
2. The operator reads the current CRDs on the management system. Once the CRDs are known to the operator, it will use a runtime client pointed to the L1 api-server to register the CRDs

#### Client Configuration
In the assisted service, we define a KUBECONFIG env variable to the service deployment that points to a mounted secret containing L1's kubeconfig. This secret is defined in a well-defined place in the management cluster and is also used by ACM. In this way, the deployed service to namespace <b>k</b> communicates with its corresponding hypershift hosted cluster out of the box with no code changes required.

> Open Issue: How and where is that secret provisioned and become known to us (AI/ACM integration issue)

Author: /cc: @jnpacker

> Open Issue: How and where is that secret provisioned and become known to us (AI/ACM integration issue)

Reviewer 3: MCE's installer will be responsible for these details. The values passed are still being defined/negotiated.

> Open Issue: How and where is that secret provisioned and become known to us (AI/ACM integration issue)

Reviewer 4: Thanks @jnpacker . Can you loop us in to that discussion?

> Open Issue: How and where is that secret provisioned and become known to us (AI/ACM integration issue)

Reviewer 3: https://issues.redhat.com/browse/ACM-1566

> ### HypershiftServerConfig controller (Infrastructure operator)

Author: Indeed. The correct name is HypershiftAgentServiceConfig.

#### Client configuration
EXTERNAL_KUBECONFIG  env variable to the service deployment points to a mounted secret containing L1's kubeconfig. This secret is defined in a well-defined place in the management cluster and is also used by ACM.

> The controller shall register itself to watch and manage resources with both L0 and L1 api-servers and instantiate 2 clients, one for each control plane.

Reviewer 2: I thought we couldn't actually watch resources from our operator in L1.

Author: We must configure the CRDs, endpoint, services and RBAC there.

#### Separating L0 and L1 resources
The resources to watch will be created by a factory and divided into L0 and L1 resources. 

The infrastructure controller then uses the appropriate clients and the inventory of resources to provision and update the necessary configuration according to the scheme above.

#### RBAC

> As with the issue of CRDs, the operator needs to define the RBAC per L1 deployment. The same options for finding which resources we need to deploy:

Author: TODO: Keep a specific list of what we need for L0

1. Build mechanism takes the RBAC from the deploy yaml and stores them on the filesystem of the operator
2. The operator reads the current RBAC definition on the management system

Option 1 is the selected one.

### Webhooks
Reference: [Dynamic Admission Control](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/)
Admission webhooks are HTTP callbacks that receive admission requests and do something with them. You can define two types of admission webhooks, validating admission webhook and mutating admission webhook. To register admission webhooks, create MutatingWebhookConfiguration or ValidatingWebhookConfiguration API objects.
These configuration objects need to be registered with L1 api-server.

The admission logic is served by an admission webhook server which serves as an extension to the api server. The pod that host the webhook server is deployed on the management cluster (L0).

In our current assisted-service deployment the communication between the api-server and the webhook server is done by a service since both runs on the same cluster. In the hypershift case, it should be done via konnectivity agent and client.

> To accomplish that we need to:

Reviewer 4: Could we just expose the webhook with a Route on L0, then on L1 refer to that route's URL when registering the webhook?

Reviewer 5: I think there is an egress configuration in L1 kube-apiserver based on the design of hypershift. All traffic of L1 kube-apiserver will go to konnectivity-server.

And konnectivity-server only can proxy traffic to konnectivity-agent.  So a Route may not help in this case.

This option was rejected because it requires too much behavioral branching in the current implementation and requires taking care of edge cases such as detecting when the user deployed multiple CRDs in a non-hypershift environment and rejecting the such option. Also, having the namespace field in the Spec is a hack.

#### Selected solution
<b>New CRD HypershiftServiceConfig on a namespace scope<b>
- Separation of intent
- Natural namespacing

### Assisted-service controllers working with hostedCluster (L1) control plane
From the user point of view, they define the cluster installation CRs on one of the namespaces in the hypershift cluster, and the assisted service picks them up and spins the cluster. To achieve that, we need to:
1. Have all cluster deployment CRDs defined on L1 
2. Assisted service's controllers should work with L1 api-server (the hosted cluster) exclusively, even though the pods on which the controllers are running are hosted on L0 (the management cluster).

#### CRD deployment
This task, which is currently done by the OLM on the management cluster, should now be carried out by the operator.

> The HASC controller on L0 is going to list all the assisted-service-related CRDs currently defined on the L0 cluster. Those CRDs are placed there by the OLM itself. So in this way, we are ensuring compatibility with the non-hypershift installation. 

Author: What you are suggesting is exactly the way we are implementing it. Which means that the above section is not clear enough? I will try to rephrase it.

> The HASC controller on L0 is going to list all the assisted-service-related CRDs currently defined on the L0 cluster. Those CRDs are placed there by the OLM itself. So in this way, we are ensuring compatibility with the non-hypershift installation. 

Author: I put some of your wording from your comment into the above paragraph .

The controller shall register itself to watch and manage resources with both L0 and L1 api-servers and shall instantiate 2 clients, one for each control plane.

Open Issue: How and where is that secret provisioned and become known to us (AI/ACM integration issue)

#### Separating L0 and L1 resources
The resources to watch will be created by a factory and divided into L0 and L1 resources. 

The infrastructure controller then uses the appropriate clients and the inventory of resources to provision and update the necessary configuration according to the scheme above.

#### RBAC
On L1 we need the following permissions:
- To support admission webhook configuration, add a ClusterRole to access validating/mutating webhook configurations, subjectaccessreviews
- To support the service's controllers, add a ClusterRole to access secrets, routes, nodes, pods, endpoints, services, etc.

This is a mirroring of the RBAC state on L0 in non-hypershift setup.

On L0 we need a few permission that previously were defined by the OLM to allow the service to operate:
- To support the service's leader election mechanism a Role to access leases and configmaps

> Open Issue: Can the leader-election-related roles be defined only on L1?

Author: We still need to verify it. In the mean time it is an open issue

> Open Issue: Can the leader-election-related roles be defined only on L1?

Reviewer 4: I'm confident it will work, but testing to verify is of course reasonable. I wrote one of the two common leader election implementations and am very familiar with the other.

The admission logic is served by an admission webhook server which serves as an extension to the api server. The pod that host the webhook server is deployed on the management cluster (L0).

In our current assisted-service deployment the communication between the api-server and the webhook server is done by a service since both runs on the same cluster. In the hypershift case, it should be done via konnectivity agent and client.

To accomplish that we need to:
1. On L0: Deploy a service that exposes the webhook admission server
2. On L1: Define a headless service with endpoint that points to the L0 service's ClusterIP
3. Specify the L1 service as the connection point to the webhook on the webhook configuration

Konnectivity server on the hostedCluster (hypershift) will forward the communication to the agent on the
management cluster and from there using the L0 service it will be forwarded to the admision pod.

RBAC definition for the webhook need now be defined on both L0 and L1 (where in the central implementation they were defined only on L0)

## Implementation Details
### HypershiftAgentClusterConfig CRD
```
type HypershiftAgentServiceConfigSpec struct {

> 	AgentServiceConfigSpec <-- re-using and keeping track of the configuration details of AgentClusterConfig

Reviewer 4: We probably just want to embed this resource, rather than make it a field and sub-object.

Reviewer 4: LGTM. My recollection is that it looked different when I reviewed it, but maybe I just misinterpreted.
